bootstrap/kubeadm/internal/cloudinit/controlplane_join.go

```diff
@@ -18,11 +18,132 @@ package cloudinit
 
 import (
 	"github.com/pkg/errors"
-
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3"
 	"sigs.k8s.io/cluster-api/util/secret"
 )
 
 const (
+	controlPlaneJoinScript = `#!/bin/bash
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Log an error and exit.
+# Args:
+#   $1 Message to log with the error
+#   $2 The error code to return
+log::error_exit() {
+	local message="${1}"
+	local code="${2}"
+
+	log::error "${message}"
+	log::info "Removing member from cluster status"
+	kubeadm reset -f update-cluster-status || true
+	log::info "Removing etcd member"
+	kubeadm reset -f remove-etcd-member || true
+	log::info "Resetting kubeadm"
+	kubeadm reset -f || true
```

Comment:
what is the reason for invoking phases and then doing full reset?

Reply:
Noticed that in some failure modes, kubeadm doesn't know it set up etcd, and then a `kubeadm reset` doesn't rewind it. Will this work to force it, or would it always be broken?